File: frontend/src/Components/Popups/UpdateDashboardItemPopup.js
import React, { useEffect, useState } from "react";
import server from "../Static/Constants";

const UpdateDashboardItemPopup = ({ dashboardSerial, dashboardItemSerial, onClose }) => {
  const [itemType, setItemType] = useState("");
  const [itemOrder, setItemOrder] = useState("");
  const [dataItemName, setDataItemName] = useState("");

  const [graphType, setGraphType] = useState("");
  const [graphTypeOptions, setGraphTypeOptions] = useState([]);
  const [columns, setColumns] = useState([]);
  const [column1, setColumn1] = useState("");
  const [column2, setColumn2] = useState("");
  const [graphTitle, setGraphTitle] = useState("");
  const [xAxisTitle, setXAxisTitle] = useState("");
  const [yAxisTitle, setYAxisTitle] = useState("");

  const [dataLines, setDataLines] = useState([]);
  const [textHeader, setTextHeader] = useState("");
  const [textBody, setTextBody] = useState("");

  const [error, setError] = useState("");

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch(`${server}/get/dashboard_item_data/${dashboardSerial}/${dashboardItemSerial}/`, {
          headers: {
            Authorization: localStorage.getItem("token"),
          },
        });

        if (response.ok) {
          const { data } = await response.json();
          setItemType(data.item_type);
          setItemOrder(data.item_order);
          setDataItemName(data.data_item_name || "");

          if (data.item_type === "Graph") {
            setGraphTypeOptions(data.graph_type || []);
            setGraphType((data.current_graph_type || "").toLowerCase());
            setColumn1(data.column_1 || "");
            setColumn2(data.column_2 || "");
            setGraphTitle(data.graph_title || "");
            setXAxisTitle(data.x_axis_title || "");
            setYAxisTitle(data.y_axis_title || "");
          } else if (data.item_type === "Table") {
            setDataLines(data.data_lines || []);
          } else if (data.item_type === "Text") {
            setTextHeader(data.text_header || "");
            setTextBody(data.text_body || "");
          }
        } else {
          const errorData = await response.json();
          setError(errorData.message || "Failed to fetch dashboard item.");
        }
      } catch (err) {
        console.error("Error fetching dashboard item:", err);
        setError("Unexpected error occurred.");
      }
    };

    const fetchColumnMetadata = async () => {
      try {
        const response = await fetch(`${server}/get/data_source_detials/${dashboardSerial}/`, {
          method: "GET",
          headers: {
            Authorization: localStorage.getItem("token"),
          },
        });

        if (response.ok) {
          const result = await response.json();
          const rawData = result.data || {};
          const processedColumns = Object.entries(rawData).map(([key, [name, type]]) => ({
            key,
            name,
            type,
          }));
          setColumns(processedColumns);
          fetchData();
        }
      } catch (e) {
        console.error("Failed to fetch column metadata", e);
      }
    };

    fetchColumnMetadata();
  }, [dashboardSerial, dashboardItemSerial]);

  const isNumeric = (colName) => {
    const column = columns.find((col) => col.name === colName);
    return column && (column.type === "int64" || column.type === "float64");
  };

  const updateLine = (index, field, value) => {
    const newLines = [...dataLines];
    newLines[index][field] = value;

    if (field === "source_1" && !isNumeric(value)) {
      newLines[index].source_2 = "";
      newLines[index].operation = "";
    }

    setDataLines(newLines);
  };

  const handleSubmitUpdate = async () => {
    const basePayload = {
      item_type: itemType,
      item_order: itemOrder,
      data_item_name: dataItemName,
      data_item_type: itemType,
      data_item_description: "",
    };

    let extraPayload = {};

    if (itemType === "Graph") {
      extraPayload = {
        graph_type: graphType.toLowerCase(),
        columns: columns.map((c) => c.name),
        column_1: column1,
        column_2: column2,
        graph_title: graphTitle,
        x_axis_title: xAxisTitle,
        y_axis_title: yAxisTitle,
      };
    } else if (itemType === "Table") {
      extraPayload = {
        data_lines: dataLines.map((line, index) => ({
          serial: line.serial || index,
          column_order: Number(line.column_order || index),
          column_name: line.column_name || "",
          source_1: line.source_1 || "",
          source_2: line.source_2 || "",
          operation: line.operation || "",
        })),
      };
    } else if (itemType === "Text") {
      extraPayload = {
        text_header: textHeader,
        text_body: textBody,
      };
    }

    try {
      const response = await fetch(`${server}/update/dashboard_item/${dashboardSerial}/${dashboardItemSerial}/`, {
        method: "PATCH",
        headers: {
          Authorization: localStorage.getItem("token"),
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ ...basePayload, ...extraPayload }),
      });

      if (response.ok) {
        alert("Dashboard item updated successfully!");
        onClose();
      } else {
        const errorData = await response.json();
        setError(errorData.message || "Update failed.");
      }
    } catch (e) {
      console.error("Update error", e);
      setError("Unexpected error occurred.");
    }
  };

  return (
    <div className="analytics-popup">
      <div className="analytics-popup-inner">
        <h3>Update Dashboard Item</h3>

        {error && <p className="analytics-popup-error">{error}</p>}

        <label>Item Order</label>
        <input type="number" value={itemOrder} onChange={(e) => setItemOrder(e.target.value)} />

        <label>Data Item Name</label>
        <input type="text" value={dataItemName} onChange={(e) => setDataItemName(e.target.value)} />

        {itemType === "Graph" && (
          <>
            <label>Graph Type</label>
            <select value={graphType} onChange={(e) => setGraphType(e.target.value)}>
              <option value="">--Select Graph Type--</option>
              {graphTypeOptions.map((type) => (
                <option key={type} value={type.toLowerCase()}>
                  {type}
                </option>
              ))}
            </select>

            <label>Column 1</label>
            <select value={column1} onChange={(e) => setColumn1(e.target.value)}>
              <option value="">--Select Column--</option>
              {columns.map((col) => (
                <option key={col.key} value={col.name}>
                  {col.name} ({col.type})
                </option>
              ))}
            </select>

            <label>Column 2</label>
            <select value={column2} onChange={(e) => setColumn2(e.target.value)}>
              <option value="">--Select Column--</option>
              {columns.map((col) => (
                <option key={col.key} value={col.name}>
                  {col.name} ({col.type})
                </option>
              ))}
            </select>

            <label>Graph Title</label>
            <input type="text" value={graphTitle} onChange={(e) => setGraphTitle(e.target.value)} />

            <label>X Axis Title</label>
            <input type="text" value={xAxisTitle} onChange={(e) => setXAxisTitle(e.target.value)} />

            <label>Y Axis Title</label>
            <input type="text" value={yAxisTitle} onChange={(e) => setYAxisTitle(e.target.value)} />
          </>
        )}

        {itemType === "Table" && (
          <>
            <label>Data Lines</label>
            {dataLines.map((line, idx) => {
              const isNum = isNumeric(line.source_1);
              return (
                <div key={idx} className="analytics-popup-line">
                  <input
                    type="number"
                    placeholder="Order"
                    value={line.column_order}
                    onChange={(e) => updateLine(idx, "column_order", e.target.value)}
                  />
                  <input
                    placeholder="Column Name"
                    value={line.column_name}
                    onChange={(e) => updateLine(idx, "column_name", e.target.value)}
                  />
                  <select
                    value={line.source_1}
                    onChange={(e) => updateLine(idx, "source_1", e.target.value)}
                  >
                    <option value="">-- Source 1 --</option>
                    {columns.map((col) => (
                      <option key={col.name} value={col.name}>
                        {col.name}
                      </option>
                    ))}
                  </select>
                  <select
                    value={line.source_2}
                    onChange={(e) => updateLine(idx, "source_2", e.target.value)}
                    disabled={!isNum}
                  >
                    <option value="">-- Source 2 --</option>
                    {columns
                      .filter((c) => isNumeric(c.name))
                      .map((col) => (
                        <option key={col.name} value={col.name}>
                          {col.name}
                        </option>
                      ))}
                  </select>
                  <select
                    value={line.operation}
                    onChange={(e) => updateLine(idx, "operation", e.target.value)}
                    disabled={!isNum}
                  >
                    <option value="">-- Operation --</option>
                    <option value="add">Add</option>
                    <option value="subtract">Subtract</option>
                    <option value="multiply">Multiply</option>
                    <option value="divide">Divide</option>
                  </select>
                  <button onClick={() => setDataLines(dataLines.filter((_, i) => i !== idx))}>❌</button>
                </div>
              );
            })}
            <button
              onClick={() =>
                setDataLines([
                  ...dataLines,
                  { column_order: dataLines.length + 1, column_name: "", source_1: "", source_2: "", operation: "" },
                ])
              }
            >
              Add Line
            </button>
          </>
        )}

        {itemType === "Text" && (
          <>
            <label>Text Header</label>
            <input type="text" value={textHeader} onChange={(e) => setTextHeader(e.target.value)} />

            <label>Text Body</label>
            <textarea rows="4" value={textBody} onChange={(e) => setTextBody(e.target.value)} />
          </>
        )}

        <div className="analytics-popup-buttons">
          <button onClick={handleSubmitUpdate}>Update</button>
          <button onClick={onClose}>Cancel</button>
        </div>
      </div>
    </div>
  );
};

export default UpdateDashboardItemPopup;

File: flutter_frontend/lib/assets/analytics/popups/updateDashboardItemPopup.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_frontend/static/constants.dart';

class UpdateDashboardItemPopup extends StatefulWidget {
  final String dashboardSerial;
  final String dashboardItemSerial;
  final VoidCallback onClose;

  const UpdateDashboardItemPopup({
    super.key,
    required this.dashboardSerial,
    required this.dashboardItemSerial,
    required this.onClose,
  });

  @override
  State<UpdateDashboardItemPopup> createState() =>
      _UpdateDashboardItemPopupState();
}

class _UpdateDashboardItemPopupState extends State<UpdateDashboardItemPopup> {
  final _storage = const FlutterSecureStorage();

  bool loading = true;
  String error = '';

  String itemType = '';
  String graphType = '';

  List<Map<String, String>> columns = [];
  List<String> graphTypeOptions = [];
  List<Map<String, dynamic>> dataLines = [];

  final itemOrderCtrl = TextEditingController();
  final dataItemNameCtrl = TextEditingController();
  final graphTitleCtrl = TextEditingController();
  final xAxisTitleCtrl = TextEditingController();
  final yAxisTitleCtrl = TextEditingController();
  final textHeaderCtrl = TextEditingController();
  final textBodyCtrl = TextEditingController();

  String column1 = '';
  String column2 = '';

  static const operations = ['add', 'subtract', 'multiply', 'divide'];

  @override
  void initState() {
    super.initState();
    _load();
  }

  @override
  void dispose() {
    itemOrderCtrl.dispose();
    dataItemNameCtrl.dispose();
    graphTitleCtrl.dispose();
    xAxisTitleCtrl.dispose();
    yAxisTitleCtrl.dispose();
    textHeaderCtrl.dispose();
    textBodyCtrl.dispose();
    super.dispose();
  }

  Future<void> _load() async {
    await _fetchColumns();
    await _fetchItem();
    setState(() => loading = false);
  }

  Future<void> _fetchColumns() async {
    final token = await _storage.read(key: 'token');
    final res = await http.get(
      Uri.parse('$server/get/data_source_detials/${widget.dashboardSerial}/'),
      headers: {'Authorization': token ?? ''},
    );

    final decoded = jsonDecode(res.body)['data'] ?? {};
    columns = decoded.entries.map<Map<String, String>>((e) {
      final List v = e.value;
      return {
        'key': e.key.toString(),
        'name': v[0].toString(),
        'type': v[1].toString(),
      };
    }).toList();
  }

  Future<void> _fetchItem() async {
    final token = await _storage.read(key: 'token');
    final res = await http.get(
      Uri.parse(
        '$server/get/dashboard_item_data/${widget.dashboardSerial}/${widget.dashboardItemSerial}/',
      ),
      headers: {'Authorization': token ?? ''},
    );

    final data = jsonDecode(res.body)['data'];

    itemType = data['item_type'];
    itemOrderCtrl.text = data['item_order'].toString();
    dataItemNameCtrl.text = data['data_item_name'] ?? '';

    if (itemType == 'Graph') {
      graphTypeOptions = List<String>.from(data['graph_type'] ?? []);
      graphType = data['current_graph_type'] ?? '';
      column1 = data['column_1'] ?? '';
      column2 = data['column_2'] ?? '';
      graphTitleCtrl.text = data['graph_title'] ?? '';
      xAxisTitleCtrl.text = data['x_axis_title'] ?? '';
      yAxisTitleCtrl.text = data['y_axis_title'] ?? '';
    }

    if (itemType == 'Table') {
      dataLines = List<Map<String, dynamic>>.from(data['data_lines'] ?? []);
    }

    if (itemType == 'Text') {
      textHeaderCtrl.text = data['text_header'] ?? '';
      textBodyCtrl.text = data['text_body'] ?? '';
    }
  }

  bool isNumeric(String? name) {
    if (name == null) return false;
    final col = columns.firstWhere((c) => c['name'] == name, orElse: () => {});
    return col['type'] == 'int64' || col['type'] == 'float64';
  }

  @override
  Widget build(BuildContext context) {
    return Dialog(
      insetPadding: const EdgeInsets.all(24),
      child: SizedBox(
        width: 1100,
        child: loading
            ? const Padding(
                padding: EdgeInsets.all(32),
                child: Center(child: CircularProgressIndicator()),
              )
            : SingleChildScrollView(
                padding: const EdgeInsets.all(24),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text(
                      'Update Dashboard Item',
                      style: TextStyle(
                        fontSize: 20,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    const SizedBox(height: 20),
                    _commonFields(),
                    const SizedBox(height: 24),
                    if (itemType == 'Graph') _graphSection(),
                    if (itemType == 'Table') _tableSection(),
                    if (itemType == 'Text') _textSection(),
                    const SizedBox(height: 32),
                    _actions(),
                  ],
                ),
              ),
      ),
    );
  }

  Widget _commonFields() => Column(
    children: [
      TextField(
        controller: itemOrderCtrl,
        decoration: const InputDecoration(labelText: 'Item Order'),
        keyboardType: TextInputType.number,
      ),
      const SizedBox(height: 12),
      TextField(
        controller: dataItemNameCtrl,
        decoration: const InputDecoration(labelText: 'Data Item Name'),
      ),
    ],
  );

  Widget _graphSection() => Column(
    crossAxisAlignment: CrossAxisAlignment.start,
    children: [
      _sectionTitle('Graph Settings'),
      _dropdown(
        label: 'Graph Type',
        value: graphTypeOptions.contains(graphType) ? graphType : null,
        items: graphTypeOptions,
        onChanged: (v) => setState(() => graphType = v ?? ''),
      ),
      _dropdown(
        label: 'Column 1',
        value: column1.isNotEmpty ? column1 : null,
        items: columns.map((c) => c['name']).whereType<String>().toList(),
        onChanged: (v) => setState(() => column1 = v ?? ''),
      ),
      _dropdown(
        label: 'Column 2',
        value: column2.isNotEmpty ? column2 : null,
        items: columns.map((c) => c['name']).whereType<String>().toList(),
        onChanged: (v) => setState(() => column2 = v ?? ''),
      ),
      TextField(
        controller: graphTitleCtrl,
        decoration: const InputDecoration(labelText: 'Graph Title'),
      ),
      TextField(
        controller: xAxisTitleCtrl,
        decoration: const InputDecoration(labelText: 'X Axis Title'),
      ),
      TextField(
        controller: yAxisTitleCtrl,
        decoration: const InputDecoration(labelText: 'Y Axis Title'),
      ),
    ],
  );

  Widget _tableSection() {
    final List<String> colNames = columns
        .map((c) => c['name'])
        .whereType<String>()
        .toList();

    final List<String> numericColNames = columns
        .where((c) => isNumeric(c['name']))
        .map((c) => c['name'])
        .whereType<String>()
        .toList();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        _sectionTitle('Table Columns'),
        SingleChildScrollView(
          scrollDirection: Axis.horizontal,
          child: Column(
            children: [
              for (int i = 0; i < dataLines.length; i++)
                Padding(
                  padding: const EdgeInsets.symmetric(vertical: 6),
                  child: Row(
                    children: [
                      _fixedText(
                        label: 'Order',
                        value: dataLines[i]['column_order'],
                        width: 80,
                        onChanged: (v) =>
                            setState(() => dataLines[i]['column_order'] = v),
                      ),
                      _fixedText(
                        label: 'Name',
                        value: dataLines[i]['column_name'],
                        width: 160,
                        onChanged: (v) =>
                            setState(() => dataLines[i]['column_name'] = v),
                      ),
                      _fixedDropdown(
                        label: 'Source 1',
                        value: colNames.contains(dataLines[i]['source_1'])
                            ? dataLines[i]['source_1']
                            : null,
                        items: colNames,
                        width: 200,
                        onChanged: (v) => setState(() {
                          dataLines[i]['source_1'] = v ?? '';
                          if (!isNumeric(v)) {
                            dataLines[i]['source_2'] = '';
                            dataLines[i]['operation'] = '';
                          }
                        }),
                      ),
                      _fixedDropdown(
                        label: 'Source 2',
                        value:
                            numericColNames.contains(dataLines[i]['source_2'])
                            ? dataLines[i]['source_2']
                            : null,
                        items: numericColNames,
                        width: 200,
                        enabled: isNumeric(dataLines[i]['source_1']),
                        onChanged: (v) =>
                            setState(() => dataLines[i]['source_2'] = v ?? ''),
                      ),
                      _fixedDropdown(
                        label: 'Op',
                        value: operations.contains(dataLines[i]['operation'])
                            ? dataLines[i]['operation']
                            : null,
                        items: operations,
                        width: 140,
                        enabled: isNumeric(dataLines[i]['source_1']),
                        onChanged: (v) =>
                            setState(() => dataLines[i]['operation'] = v ?? ''),
                      ),
                      IconButton(
                        icon: const Icon(Icons.delete),
                        onPressed: () => setState(() => dataLines.removeAt(i)),
                      ),
                    ],
                  ),
                ),
            ],
          ),
        ),
        const SizedBox(height: 12),

        ElevatedButton.icon(
          icon: const Icon(Icons.add),
          label: const Text('Add Line'),
          onPressed: () {
            setState(() {
              dataLines.add({
                'column_order': dataLines.length + 1,
                'column_name': '',
                'source_1': '',
                'source_2': '',
                'operation': '',
              });
            });
          },
        ),
      ],
    );
  }

  Widget _textSection() => Column(
    crossAxisAlignment: CrossAxisAlignment.start,
    children: [
      _sectionTitle('Text Content'),
      TextField(
        controller: textHeaderCtrl,
        decoration: const InputDecoration(labelText: 'Header'),
      ),
      const SizedBox(height: 12),
      TextField(
        controller: textBodyCtrl,
        maxLines: 5,
        decoration: const InputDecoration(labelText: 'Body'),
      ),
    ],
  );

  Widget _sectionTitle(String t) => Padding(
    padding: const EdgeInsets.symmetric(vertical: 12),
    child: Text(
      t,
      style: const TextStyle(fontSize: 16, fontWeight: FontWeight.w600),
    ),
  );

  Widget _dropdown({
    required String label,
    required String? value,
    required List<String> items,
    required Function(String?) onChanged,
  }) => DropdownButtonFormField<String>(
    value: value,
    decoration: InputDecoration(labelText: label),
    items: items
        .map((v) => DropdownMenuItem(value: v, child: Text(v)))
        .toList(),
    onChanged: onChanged,
  );

  Widget _fixedText({
    required String label,
    required dynamic value,
    required double width,
    required Function(String) onChanged,
  }) => SizedBox(
    width: width,
    child: TextField(
      controller: TextEditingController(text: value?.toString() ?? ''),
      onChanged: onChanged,
      decoration: InputDecoration(labelText: label),
    ),
  );

  Widget _fixedDropdown({
    required String label,
    required String? value,
    required List<String> items,
    required double width,
    required Function(String?) onChanged,
    bool enabled = true,
  }) => SizedBox(
    width: width,
    child: DropdownButtonFormField<String>(
      value: value,
      decoration: InputDecoration(labelText: label),
      items: items
          .map((v) => DropdownMenuItem(value: v, child: Text(v)))
          .toList(),
      onChanged: enabled ? onChanged : null,
    ),
  );

  Widget _actions() => Row(
    mainAxisAlignment: MainAxisAlignment.end,
    children: [
      TextButton(onPressed: widget.onClose, child: const Text('Cancel')),
      const SizedBox(width: 12),
      ElevatedButton(onPressed: () {}, child: const Text('Update')),
    ],
  );
}
